### ISSUE TITLE
[v3.0] Use profile integer ids consistently in routes

### ISSUE THREAD

**Issue by trwnh** [tootsuite/mastodon#10745]
### Pitch
Mastodon currently assigns each user (both local and remote) an internal ID, but this is *only* exposed over the API. I propose using this ID for more routes because it is more advantageous and would allow consistency and features that aren't currently possible.

### Current behavior

Accounts are queried by ID via the API like so: https://mastodon.social/api/v1/accounts/14715

key | value
--- | ---
id | "14715"
username | "trwnh"
acct | "trwnh"
display_name | "infinite love ⴳ"
locked | false
bot | false
created_at | "2016-11-24T10:02:12.085Z"
note | "<p>i have approximate knowledge of many things. perpetual student. (nb/ace/they) <a href=\"https://mastodon.social/tags/soulpunx\" class=\"mention hashtag\" rel=\"tag\">#<span>soulpunx</span></a> <a href=\"https://mastodon.social/tags/safecamp\" class=\"mention hashtag\" rel=\"tag\">#<span>safecamp</span></a> <a href=\"https://mastodon.social/tags/solarchild\" class=\"mention hashtag\" rel=\"tag\">#<span>solarchild</span></a></p><p>xmpp/email: a@trwnh.com<br /><a href=\"https://trwnh.com\" rel=\"nofollow noopener\" target=\"_blank\"><span class=\"invisible\">https://</span><span class=\"\">trwnh.com</span><span class=\"invisible\"></span></a></p>"
url | "https://mastodon.social/@trwnh"
avatar | "https://files.mastodon.social/accounts/avatars/000/014/715/original/eadf7279b61f9be9.png"
avatar_static | "https://files.mastodon.social/accounts/avatars/000/014/715/original/eadf7279b61f9be9.png"
header | "https://files.mastodon.social/accounts/headers/000/014/715/original/4f42c68d31fc7194.jpg"
header_static | "https://files.mastodon.social/accounts/headers/000/014/715/original/4f42c68d31fc7194.jpg"

```

**Comment by ThibG** [comment]
From what I understand, you suggest using the internal account identifier for all URLs in Mastodon. This would indeed have the advantage of allowing the username part to change. However, I fail to see how it would help with changing instances.

Also, the current ActivityPub URIs already contain the username, so your proposal would… break them… there isn't really a possible migration path, the URI is not supposed to change.

Furthermore, Mastodon is weird in that it uses username + domain extensively instead (or in addition to) the URI to identify remote accounts. Which means an account changing username without changing URI will break a lot of things in current Mastodon instances. And changing that is very hard.

**Comment by trwnh** [comment]
>I fail to see how it would help with changing instances.
It helps with changing instances because the ID remains constant not only regardless of username, but also regardless of domain. Or rather, it allows the `acct` value to change, if logic is instead performed with Masto API's `id` instead of `acct`. Once all logic depends on `id`, the entire `acct` can change freely without affecting any logic. In such a case, moving to a new domain requires only copying over the data and updating the `acct` field to point to the new location. A username change on the same domain requires no copying.

>the current ActivityPub URIs already contain the username, so your proposal would… break them… 

Yes, but also, the longer this doesn't change, the more URIs there will be to break. I pointed out this issue a year ago, but the move to v3.0 presents an opportunity to pitch breaking changes. Current URIs can either be maintained for legacy compatibility, or regenerated so that future fetches grab the new `id`. This also has implications for content caching/expiry -- currently, I think accounts/statuses are cached until purged by a rake task, right?

>Mastodon is weird in that it uses username + domain extensively instead (or in addition to) the URI to identify remote accounts. Which means an account changing username without changing URI will break a lot of things in current Mastodon instances. And changing that is very hard.

And it will get harder as more time goes on. Mastodon has already had to deal with many different bugs due to this incorrect implementation, and as new ActivityPub software is written without these assumptions, Mastodon will only grow more incompatible. Put simply, the current URI's reliance on a username is an anti-pattern that persists largely because of Mastodon's existing installation base. Pleroma uses `/notice/id` and `/users/id` for its URIs, Pixelfed is planning to go back to `/p/id` soon to allow username changes by v1.0, and presumably other future software will likewise do things that are inconvenient for current Mastodon logic.

**Comment by ThibG** [comment]
> It helps with changing instances because the ID remains constant not only regardless of username, but also regardless of domain.

No, it does not? The id is local to the instance, I do not understand your point. Furthermore, the URI still needs to be referenced, so it will include the domain name, and still break on domain change?

> Yes, but also, the longer this doesn't change, the more URIs there will be to break.

By now, the protocol and network are pretty well-established, I don't think it's a case of “the earlier the change, the better”. Instead, if we do make a breaking change, we would want to be pretty sure it's worth it.

**Comment by trwnh** [comment]
>The id is local to the instance, I do not understand your point. Furthermore, the URI still needs to be referenced, so it will include the domain name, and still break on domain change?

I gave an example of how this could work. The API response would update like so:
- `id: 14715` (unchanged)
- `acct: trwnh` to `acct: a@trwnh.com`
- `url: mastodon.social/@trwnh` to `url: trwnh.com/@a`

The proposal is to stop querying/assigning `mastodon.social/users/trwnh` and to start querying/assigning `mastodon.social/users/14715`. Crucially, this works because *both local and remote users are assigned a local ID.* If such a major change were to be adopted, it would stop mattering which domain a user is actually on -- internally it's the same reference, and that's all that matters. It opens a lot of doors to things like profiles changing names, profiles being cloned to multiple servers, nomadic identity, etc. Without this change, Mastodon will not be robust enough to handle remote servers that can/will change representations at any time.. This has already been an issue in the past, and it will be a bigger issue in the future.

To reiterate, here are the current issues that Mastodon has had:
- Username cannot change at all, so Krita -> krita was a breaking change and caused Mastodon servers to stop receiving statuses until a manual fix was merged/deployed. The "fix" was to make usernames case-insensitive, so some existing users had their accounts deleted -- e.g. `@claudia@mastodon.social` was deleted because it conflicted with the also-existing `@Claudia@mastodon.social`. This could have conceivably happened to more users, given 2 years of prior account history. The fix is also not robust enough to handle username changes that aren't simply a case swap.
- Username lookup is still quite broken anyway, especially due to weirdness of constraints around symbols. Consider `@eugen_rochko@social.puri.sm` -- you would presume a URL of https://social.puri.sm/@eugen_rochko but that actually 404s! The correct URL is https://social.puri.sm/@eugen.rochko instead.
- When a domain goes down or messes up due to some error or config change, it wreaks havoc in all kinds of ways. a.weirder.earth had to rename itself to weirder.earth; mastodon.rocks was never the same after the database drop; social.lou.lt moving to lou.lt meant that accounts just suddenly stopped updating and couldn't really be refollowed; and so on. 
- Moving and account migration has been largely more complicated than it needs to be, because there simply hasn't been a way to consistently refer to the same information across domains... At best, Mastodon's migration strategy seems to be "create an entirely new account/profile and try to import anything that hasn't changed". This means maybe you can keep the list of people you were following (and refollow them), but good luck bringing your old posts/media with you. If you do, they're basically new objects entirely.

I could go on, but hopefully it should be obvious why dealing with this is "worth it".

**Comment by ThibG** [comment]
I think you have a deep misunderstanding of what the numeric identifier is. It's not global. It's local to the instance. It depends on the order in which accounts were created/discovered. That's why it's used internally.

> > The id is local to the instance, I do not understand your point. Furthermore, the URI still needs to be referenced, so it will include the domain name, and still break on domain change?
> 
> I gave an example of how this could work. The API response would update like so:
> 
>     * `id: 14715` (unchanged)
> 
>     * `acct: trwnh` to `acct: a@trwnh.com`
> 
>     * `url: mastodon.social/@trwnh` to `url: trwnh.com/@a` (or whatever activitypub `url` is, tbh)

What API are you talking about? The REST API? If so that's completely orthogonal to your proposal of using the numeric id in URLs.

> The proposal is to stop querying/assigning `mastodon.social/users/trwnh` and to start querying/assigning `mastodon.social/users/14715`. Crucially, this works because _both local and remote users are assigned a local ID._

They are, but it's not a unique identifier across instances. On mastodon.social, 1 is Gargron, on my instance, it's me. Those are not global identifiers.

> If such a major change were to be adopted, it would stop mattering which domain a user is actually on -- internally it's the same reference, and that's all that matters.

It is *not*. And even if it were, you'd still need to know where the user is hosted to be able to reach them.

> […]
>     * Username lookup is still quite broken anyway, especially due to weirdness of constraints around symbols. Consider `@eugen_rochko@social.puri.sm` -- you would presume a URL of https://social.puri.sm/@eugen_rochko but that actually 404s! The correct URL is https://social.puri.sm/@eugen.rochko instead.

This does not matter, because you do not do that lookup manually.

>     * When a domain goes down or messes up due to some error or config change, it wreaks havoc in all kinds of ways. a.weirder.earth had to rename itself to weirder.earth; mastodon.rocks was never the same after the database drop; social.lou.lt moving to lou.lt meant that accounts just suddenly stopped updating and couldn't really be refollowed; and so on.

I am not aware of the details, but I assume it's followers/followed info being out of sync, as well as stale keypairs. Your proposal would not help here.
It would actually make things worse here, since accounts would not be created in the same order, the id would not correspond to the same accounts anymore.

**Comment by trwnh** [comment]
>I think you have a deep misunderstanding of what the numeric identifier is. It's not global. It's local to the instance. It depends on the order in which accounts were created/discovered. That's why it's used internally.

Yes, exactly, which is why it should be used for internal logic -- its authority is derived from the server itself! The fact that the integer ID is local to the instance is *the entire point*. It does not have to be unique across instances -- it only has to be unique on the local instance. To clarify, when you say

>On mastodon.social, 1 is Gargron, on my instance, it's me. Those are not global identifiers. [...] It is not [the same reference]. And even if it were, you'd still need to know where the user is hosted to be able to reach them.

it's irrelevant, because mastodon.social is not asking your instance for an ID. It's asking itself. `mastodon.social/users/1` is the same as `mastodon.social/users/1` because that is a tautology. What `mastodon.social/users/1` gets translated to is entirely within the authority of the server running on `mastodon.social`. Again, as per my examples above, you could have `mastodon.social/users/14715` redirect to `trwnh.com/users/1`. I could then import my statuses such that `mastodon.social/users/14715/statuses/100560344258327930` continues to resolve to the same status dynamically. Whatever its current *location* happens to be, you can load it by requesting the *identifier*. 

The fundamental problem is that Mastodon is basically going through the effort of assigning local IDs and then not using them fully.

**Comment by 0xa** [comment]
I have thought about that before and it's a good idea to allow username changes, but i also fail to see how it would help the migrations themselves. And if you move account id 1312 from one instance to the other and it already has an account 1312 registered to another username, you will have to assign it a new identifier anyway. It's not different from usernames.

Also, having a public facing **sequential** identifier has added risks; notably that you can easily find the latest registered identifier, and enumerate account from 1 to the last. In this case, enumerate every account on the instance without authentication.
The solution to this is a longer, random, public identifier (an UUID maybe?) which should be kept in parallel to the internal ID.

I think this would be a good move for the API URLs, but i wouldn't change the rest; if you change username maybe you want to change the public URL too, but not break established follows. Old usernames can have a redirection, and the option to break links to the profile on purpose can be important. The profile URL should behave like webfinger lookups on this.

**Comment by trwnh** [comment]
For migrations you could have id 1312 on one instance move to id 1 on another instance. The two id fields do not have to be the same; id depends on which instance you are asking. If you ask mastodon.social for account 14715 you can either be served the local representation that mastodon.social is aware of, or you can be redirected to ask trwnh.com for account 1.

There is no requirement for the id to be sequential either; I'm simply proposing this as a first step to see where the discussion goes. 

A more realistic plan I wrote up long ago buried somewhere in the replies to #177 went loosely as follows:
1) refactor logic to allow using either masto `id` or masto `acct`.
3) add migration tasks to reassign profiles/statuses to id namespace instead of username namespace
4) wait for this code to percolate through the ecosystem as servers update
5) push user-facing ui to allow easier migrations

**Comment by trwnh** [comment]
tldr/recap at this point in the thread: the main proposal is to stop using uri in the form `/users/trwnh` and start using in the form `/users/14715`, which is the biggest blocker to true migration. `url` can rely on username, `id` should not. 

relevant/related comments and writeups:
- on "accounts" (authentication) vs "profiles" (content/identity): https://github.com/tootsuite/mastodon/issues/177#issuecomment-354634232
- comparisons to IMAP, XMPP, Friendica: https://github.com/tootsuite/mastodon/issues/177#issuecomment-378779767
- zot nomadic identity explainer: https://github.com/tootsuite/mastodon/issues/177#issuecomment-382761085
- an 8-step migration plan: https://github.com/tootsuite/mastodon/issues/177#issuecomment-399172654
- rehash on username vs id in uri: https://github.com/tootsuite/mastodon/issues/177#issuecomment-446882405
- exploring how pleroma does content addressing: https://github.com/tootsuite/mastodon/issues/177#issuecomment-477852263
- on local vs. global uniqueness, i.e. what inspired this issue to be filed: https://github.com/pixelfed/pixelfed/issues/216

**Comment by ThibG** [comment]
Again, the only benefit I can see in using the internal identifier for routes is the ability to transparently change usernames. It does not help for cross-instance migrations (in your examples, the instance of origin could just as well redirect to the existing username-based URLs, and that whole migration scheme sounds dubious to me, tbh).

The same benefit would be achievable by using a UUID, which would also avoid the enumeration issues as well as the possible confusions when an instance is reset/restored from backup etc. as I explained earlier.

However, that (small, but real) benefit is only achievable if Mastodon actually knows how to handle remote accounts changing `acct:` URI, which requires very deep changes in how Mastodon works and require potentially troublesome database migrations. Currently, Mastodon treats the `acct:` as authoritative, and, uh, hopes everything goes well with the URI.

Furthermore, with all that done, there is one big issue with this change: how would the migration work? Your proposal's purpose is to make the `acct:` mutable. But it would *also* mutate the ActivityPub ids to achieve it. If both things are mutable, how can Mastodon handle the change?

**Comment by trwnh** [comment]
After getting some sleep it occurred to me that the enumeration concern is not really part of this issue because it already exists and it's already public-facing. Anyone can query the `/api/v1/accounts` endpoint right now if they wish. So that's not really an argument against also using the same for `/users/` uri -- it's  separate issue.

Re: cross-instance migrations, I'm still not sure why/how I haven't made it clear exactly how having a stable reference would help in stable lookups. The "very deep changes and potentially troublesome database migrations" is the point of this issue.

Fundamentally, the `acct` should not be authoritative -- it is a reflection of webfinger. only the `id` can be trusted to never change. The id does not have to be globally unique if you only ever manage it locally. Of course, having a globally unique ID without relying on DNS would be *better*, but it's out of scope for this feature request -- eliminating reliance on DNS is a long-term, far-future goal. The short-term goal is to eliminate reliance on location as unchanging. To draw an analogy, if Mastodon were a person, it would be unable to recognize when someone moved houses/addresses. I am not `123 Main Street`, I am simply currently located there.

As far as ensuring that two profiles are owned by the same person, that can be done in any way you want -- you can OAuth from the new profile into the old profile, you can do stuff with `alsoKnownAs`, you can cryptographically sign stuff, whatever. 

re: mutating the ActivityPub URI schema, this would only happen once. It could be done with an `Update` or a `Move` if you *really* want everyone on the network to know about it -- just say that the Actor with id https://mastodon.social/users/trwnh now has id https://mastodon.social/users/14715, and the status at https://mastodon.social/users/trwnh/statuses/100560344258327930 is now https://mastodon.social/users/14715/statuses/100560344258327930. 

Ultimately the URI should not be based on things like authorship, publishing status, file extensions, etc -- even if you treat them as immutable, not everyone will. Mastodon could very well never support username changes without basically creating an entirely new profile, but that just means it will be more fragile when other systems change, and Mastodon should be able to handle that. Keep in mind that at the core of all these systems, we're really just doing web stuff with HTML and HTTP. The best practices of the past decades shouldn't be forgotten or ignored.

**Comment by trwnh** [comment]
More examples to illustrate:
- Multiple URI for same user (using API routes for demonstration -- under this issue's proposal, `/api/v1/accounts/` and `/users/` would be practically interchangeable):
  - https://mastodon.social/api/v1/accounts/14715
  - https://sunbeam.city/api/v1/accounts/249
  - https://linernotes.club/api/v1/accounts/256
  - https://anticapitalist.party/api/v1/accounts/2748
- Finding any user (local/remote) by local ID
  - Local: https://mastodon.social/users/14715 -> https://mastodon.social/@trwnh
  - Local (after migration): https://mastodon.social/users/14715 -> https://mastodon.social/@altrwnh
  - Remote (after migrating to another domain): https://mastodon.social/users/14715 -> https://trwnh.com/users/1 -> https://trwnh.com/@a

**Comment by trwnh** [comment]
Looking back at this after getting some sleep, I think the conversation may have been conflating two closely related but different issues:

1) That the current ActivityPub `id` URI is not amenable to username changes vs. the Mastodon API `id` local number;

2) That the Mastodon API `acct` field is used in business logic instead of the `id`.

The consequence of the former is that Mastodon cannot handle local username changes at all. The consequence of the latter is that Mastodon is unable to handle username changes in the slightest, even after already discovering a remote account (at which point Webfinger should no longer be necessary, from a purely theoretical standpoint -- just refer to users by the local ID!)

**Comment by redyoshi49q** [comment]
> it's irrelevant, because mastodon.social is not asking your instance for an ID. It's asking itself. `mastodon.social/users/1` is the same as `mastodon.social/users/1` because that is a tautology. What `mastodon.social/users/1` gets translated to is entirely within the authority of the server running on `mastodon.social`. Again, as per my examples above, you could have `mastodon.social/users/14715` redirect to `trwnh.com/users/1`. I could then import my statuses such that `mastodon.social/users/14715/statuses/100560344258327930` continues to resolve to the same status dynamically. Whatever its current _location_ happens to be, you can load it by requesting the _identifier_. The same object can have not only multiple URLs, but also multiple URIs. The trick is in gaining local-to-global uniqueness by using DNS as a global uniqueness guarantee, then using the local ID within that globally-unique namespace.

If my understanding of what you say here is correct, you're looking to have Mastodon APIs and URLs/URIs use user IDs by default so that username changes and migrations of a user to a different server could both be implemented via redirects.  In effect, you're viewing URLs/URIs as pointers, including in the sense that the destination of data targeted by such URLs/URIs is mutable.

For the case of server migration, while this proposal might be beneficial as a helpful prerequisite to implementing server migration, it is not by itself sufficient.  The problem I see is that the data generated by a user on their old server will still remain on that old server indefinitely, which can be problematic, for instance, when the migration is due to the old server being shut down.  Changing URLs/URIs of users on the old servers to point to their new homes on their respective new servers is not sufficient; there **must** be a mechanism to migrate data across servers as well, or users' data will remain cemented into the servers it was originally created on even as the identities of those same users are given the ability to migrate as they wish.  Simply having the data on the old server be referenced in the user's new account on a different server is not an adequate state to leave the data in permanently.

Other commenters on the user migration issue have voiced concern regarding how this data migration might be burdensome to the source and destination servers.  One advantage that I see in your paradigm of using URLs/URIs effectively as pointers is that it could allow for the option of rate limiting such a data migration without requiring the migrating user to wait until all of the data was transferred.  The state of having data on an old server being referenced by a user's new account elsewhere would occur, but only until all of that user's data is successfully migrated; in the meantime URLs/URLs to individual pieces of data (such as the contents of a particular toot) would point to the old server's copy before the transfer and then to the new server's copy after the transfer.  The user experience during the migration could continue to function almost the same as it would be if there was not an account transfer in progress, meaning that the existence of an in-progress server migration of a user account could be made almost entirely transparent (perhaps with the exception of a UI note that the associated data belongs to a user that's currently migrating across servers, as well as an indicator showing whether the specific data had made the transition between servers yet or not).

For the sake of my own clarity: do you intend for this change to affect the URLs that an end user would ultimately see in their web browsers and share with others, or do you intend for this change to primarily affect just the API/URLs/URIs as used within and between Mastodon instances?  This change would be less attractive if it inhibits users from being able to use URLs containing human readable elements (such as their own usernames).

**Comment by trwnh** [comment]
>In effect, you're viewing URLs/URIs as pointers [...] so that username changes and migrations of a user to a different server could both be implemented via redirects

Essentially, yeah. The end goal is to serve a certain JSON document/response, regardless of how you arrived at it. The pointers can be anything; only the originating server gets to decide the canonical route, but that **doesn't stop the local server from creating its own routes for internal referencing.** All it requires is deduplicating/updating the internal links whenever the destination changes.

>while this proposal might be beneficial as a helpful prerequisite to implementing server migration, it is not by itself sufficient

Yes. It is not sufficient; it simply makes migration easier by laying the groundwork to simply update some pointers rather than redownload / rewrite the resources themselves.

>data generated by a user on their old server will still remain on that old server indefinitely, which can be problematic, for instance, when the migration is due to the old server being shut down

Only if the old posts are not also migrated -- which would be easier if you already *have* the data, it just becomes a question of attribution (and cache in/validation, in case the actor wishes to not keep the data). **Similar to how when your friend changes email addresses, you can still keep the old emails you received, you just start sending your messages to the new address, and you don't have to have your friend resend every single email they've ever sent you in the past.**

>The user experience during the migration could continue to function almost the same as it would be if there was not an account transfer in progress, meaning that the existence of an in-progress server migration of a user account could be made almost entirely transparent

Arguably it would not only make the migration transparent, but it would also pave the way for users to have multiple addresses that could be used interchangeably as aliases. It's possible to decouple accounts from profiles from addresses from status routes -- but the first step *has* to be refactoring all the business logic to allow for this. And the easiest way to do that is to simply use the internal references that already exist, rather than deferring to the current address (which is not fully dereferenced, technically).

>For the sake of my own clarity: do you intend for this change to affect the URLs that an end user would ultimately see in their web browsers and share with others, or do you intend for this change to primarily affect just the API/URLs/URIs as used within and between Mastodon instances?

The latter, of course. `id` should be a machine-readable route (and just a database value internally), whereas `url` is the human-readable route containing the current username. `mastodon.social/@trwnh` would redirect to `mastodon.social/users/14715` which could then redirect further to `trwnh.com/u/1` which serves the JSON document for my actor. Users can continue to identify as a Webfinger address or as a profile URL, but the software should not stop there. That's what dereferencing is ultimately about: turning references into resources.

**Comment by trwnh** [comment]
welp, bumping this down the road to 5.0

**Comment by Ciantic** [comment]
People are already depending on `@something@example.com` working, e.g. [Simon Willison made an article](https://til.simonwillison.net/mastodon/custom-domain-mastodon) how to make these vanity handles.

IMO it should be made *even* easier, e.g. if John Doe had account in: https://mastodon.social/@JohnDoe

And he owns example.com, he should be allowed to make `@john@example.com` as his handle simply by creating some easy linking file e.g. `https://example.com/fediverse.json`

```jsonc
{
	"@john@example.com": {
		"aliases": ["https://mastodon.social/@JohnDoe", "https://mastodon.social/users/JohnDoe"]
	}
}
```

(Of course it would require a field in mastodon.social as well that you want to change the handle)

**Comment by trwnh** [comment]
#32724 looks like it implements this for at least new accounts

**Comment by ShadowJonathan** [comment]
> This would also make the upcoming Move and account migration simpler, as it would no longer be necessary to move accounts just to change a username. User 14715 could easily change from trwnh to a@trwnh.com with zero side effects for lookup via the API.

This is, as long as the original server keeps responding with 301s of its previous posts. Or; as long as the original server is up.

(Yes, I know i'm bumping, apologies for that)

**Comment by ShadowJonathan** [comment]
Also, #32724 has been merged :)
